src/conn_sock.c

```diff
@@ -349,27 +349,32 @@ char *socket_parent_dir(gboolean use_full_attach_path, size_t desired_len)
 	return base_path;
 }
 
-
 void schedule_main_stdin_write()
 {
 	schedule_local_sock_write(&local_mainfd_stdin);
 }
 
-void write_back_to_remote_consoles(char *buf, int len)
+static void write_remote_sock(gpointer data, G_GNUC_UNUSED gpointer user_data)
 {
-	if (local_mainfd_stdin.readers == NULL)
+	struct remote_sock_s *remote_sock = (struct remote_sock_s *)data;
+	if (!remote_sock->writable)
 		return;
-
-	for (int i = local_mainfd_stdin.readers->len; i > 0; i--) {
-		struct remote_sock_s *remote_sock = g_ptr_array_index(local_mainfd_stdin.readers, i - 1);
-
-		if (remote_sock->writable && write_all(remote_sock->fd, buf, len) < 0) {
-			nwarn("Failed to write to remote console socket");
-			remote_sock_shutdown(remote_sock, SHUT_WR);
-		}
+	writev_iov_t *iov = (writev_iov_t *)user_data;
+	if (writev_all(remote_sock->fd, iov) < 0) {
+		nwarn("Failed to write to remote console socket");
+		remote_sock_shutdown(remote_sock, SHUT_WR);
 	}
 }
 
+void write_back_to_remote_consoles(stdpipe_t pipe, char *buf, size_t buflen)
+{
+	if (local_mainfd_stdin.readers == NULL)
+		return;
+	struct iovec iov[2] = {{&pipe, 1}, {buf, buflen}};
+	writev_iov_t wviov = {2, 2, iov};
+	g_ptr_array_foreach(local_mainfd_stdin.readers, write_remote_sock, &wviov);
```

Comment:
The `wviov` is created few lines above, passed as pointer down to `write_remote_sock` for each `local_mainfd_stdin.reader`. I think the issue is that the first `writev_all` call changes the `wviov` and the second call of `writev_all` for the second reader from `local_mainfd_stdin.readers` will get a `wviov` with `buf->iovcnt = 0;`, effectively writing nothing.

I think we will have to rebuild the writev_iov_t in the `write_remote_sock` or change the `writev_all` to operate on a local copy and do not mutate it.

I think it would be also great to add a bats test for two readers, so this use-case is covered.